src/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMeanAndListModeDataWithProjMatrixByBin.cxx

```diff
@@ -312,322 +320,265 @@ PoissonLogLikelihoodWithLinearModelForMeanAndListModeDataWithProjMatrixByBin<Tar
 {
-    if(!this->recompute_cache && this->cache_lm_file)
-    {
-        info("Reading cache from disk...");
-        std::string curr_dir;
-        if (this->cache_path.size() > 0)
-            curr_dir = this->cache_path;
-        else
-            curr_dir = FilePath::get_current_working_directory();
-        std::string cache_filename = "my_CACHE00.bin";
-        FilePath icache(cache_filename, false);
-        icache.prepend_directory_name(curr_dir);
-
-        if (icache.is_regular_file())
-        {
-            info( boost::format("Loading Listmode cache from disk %1%") % icache.get_as_string());
-            std::ifstream fin(icache.get_as_string(), std::ios::in | std::ios::binary
-                              | std::ios::ate);
+    info( boost::format("Storing Cached Events ... "));
 
-            unsigned long int num_of_records = fin.tellg()/sizeof (Bin);
-            record_cache.reserve(num_of_records);
+    std::string cache_filename = "my_CACHE" + std::to_string(file_id) +".bin";
+    FilePath ocache(cache_filename, false);
+    ocache.prepend_directory_name(this->get_cache_path());
 
-            fin.clear();
-            fin.seekg(0);
+    bool with_add = !is_null_ptr(this->additive_proj_data_sptr);
 
-            while(!fin.eof())
-            {
-                BinAndCorr tmp;
-                fin.read((char*)&tmp, sizeof(Bin));
-                if (this->has_add)
-                {
-                    tmp.my_corr = tmp.my_bin.get_bin_value();
-                    tmp.my_bin.set_bin_value(1);
-                }
-                record_cache.push_back(tmp);
-            }
-            //The while will push one junk record
-            record_cache.pop_back();
-            fin.close();
-        }
-        else
+    {
+        info("Storing Listmode cache to file \"" + ocache.get_as_string() + "\".");
+        // open the file, overwriting whatever was there before
+        std::ofstream fout(ocache.get_as_string(), std::ios::out | std::ios::binary | std::ios::trunc);
+        if (!fout)
+          error("Error opening cache file \"" + ocache.get_as_string() + "\" for writing.");
+
+        for(unsigned long int ie = 0; ie < record_cache.size(); ++ie)
         {
-            error("Cannot find Listmode cache on disk. Please recompute or comment out the  max cache size. Abort.");
-            return Succeeded::no;
+            Bin tmp = record_cache[ie].my_bin;
+            if(with_add)
+              tmp.set_bin_value(record_cache[ie].my_corr);
+            fout.write((char*)&tmp, sizeof(Bin));
         }
+        if (!fout)
+          error("Error writing to cache file \"" + ocache.get_as_string() + "\".");
+
+        fout.close();
+    }
+
+    num_cache_files += 1;
+    std::cout << "Caches: " << num_cache_files << std::endl;
+    // clear
+    record_cache.clear();
+    record_cache.reserve(this->cache_size);
```

Comment:
@NikEfth why would we do this? write a cache file shouldn't affect `record_cache`

Reply:
ok, I now understand the need for the `clear` (as otherwise we might re-use previous events). Still I find it surprising this happens in a `write` function so will move it out